crates/bevy_pbr/src/material.rs

```diff
@@ -201,7 +201,6 @@ impl<M: SpecializedMaterial> Default for MaterialPlugin<M> {
 impl<M: SpecializedMaterial> Plugin for MaterialPlugin<M> {
     fn build(&self, app: &mut App) {
         app.add_asset::<M>()
-            .add_plugin(ExtractComponentPlugin::<Handle<M>>::default())
```

Comment:
why didn't you add the visibility filter in the general plugin instead of doing a custom version for just material?

Reply:
You can only filter on component presence? ComputedVisibility is a bool. When I was benchmarking using ComputedVisibility as a marker component, I would then use it as a filter. But the adding/removing of a ComputedVisibility marker component had significant cost as I recall

Reply:
I mean doing
```diff
--- a/crates/bevy_render/src/render_component.rs
+++ b/crates/bevy_render/src/render_component.rs
@@ -164,13 +164,15 @@ impl<T: Asset> ExtractComponent for Handle<T> {
 fn extract_components<C: ExtractComponent>(
     mut commands: Commands,
     mut previous_len: Local<usize>,
-    mut query: StaticSystemParam<Query<(Entity, C::Query), C::Filter>>,
+    mut query: StaticSystemParam<Query<(Entity, ComputedVisibility, C::Query), C::Filter>>,
 ) where
     <C::Filter as WorldQuery>::Fetch: FilterFetch,
 {
     let mut values = Vec::with_capacity(*previous_len);
-    for (entity, query_item) in query.iter_mut() {
-        values.push((entity, (C::extract_component(query_item),)));
+    for (entity, computed_visiblity, query_item) in query.iter_mut() {
+        if computed_visiblity.is_visible {
+            values.push((entity, (C::extract_component(query_item),)));
+        }
     }
     *previous_len = values.len();
     commands.insert_or_spawn_batch(values);
```

so that all `ExtractComponentPlugin` would check for visibility

Reply:
Then all components being extracted would have to have a ComputedVisibility component. I suppose you could use an Option<> in the query. And `if maybe_computed_visibility.is_none() or maybe_computed_visibility.unwrap().is_visible` but it also feels a bit unexpected perhaps?

Reply:
Do we have the guarantee that all `SpecializedMaterial` will have the component `ComputedVisibility`?

Reply:
> Do we have the guarantee that all `SpecializedMaterial` will have the component `ComputedVisibility`?

They would usually be constructed with `MaterialMeshBundle` which includes a `ComputedVisibility`, but we can't guarantee it I guess.

> It could make sense to use `ComputedVisibility` as a global filter on everything that will be extracted?

It makes sense to me, from a rendering perspective. Something has to be directly or indirectly (say there is a tree behind you and a light behind the tree that illuminates in front of you and the tree casts shadows - the tree is not directly visible but the shadows show you it is there) visible to influence what is seen. But I am not certain that every component that is extracted is on an entity that has a computed visibility component.

Reply:
I don't see any internal uses of ExtractComponentPlugin to extract components on an entity that does not have ComputedVisibility at least. We could make it a requirement if using ExtractComponentPlugin I suppose, and people can still write their own extraction functions for other cases.

What do you think @cart ?

Reply:
I agree that hard-coding here feels odd given how general this is. Visibility is pretty generalized, but I imagine it isn't "exactly" as general as the concept of "component extraction".  Maybe we just need an `ExtractVisibleComponentPlugin` variant? Or an `ExtractComponentPlugin::<T>::extract_visible()` constructor?

Reply:
How about a query that includes `Option<ComputedVisibility>` and extracts if `maybe_computed_visibility.map_or(true, |computed_visibility| computed_visibility.is_visible)`?

Reply:
that feels like it would make all extract a little slower when we should be able to say at build time wether there is the  `ComputedVisibility` component or not

Reply:
The alternative that I see is to have two separate queries, one with ComputedVisibility in it, one with a Without<ComputedVisibility> filter.

Reply:
yup, following Cart suggestion of using `ExtractComponentPlugin::<T>::extract_visible()` to trigger using the other query sounds nice to me